### PR TITLE
Auto set prediction_context for offline prediction

### PR DESCRIPTION
**Pull request by serena-ruan**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/serena-ruan/mlflow/pull/12084?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/12084/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 12084
```

</p>
</details>

### Related Issues/PRs

 #xxx

### What changes are proposed in this pull request?
We want to align online predict and offline predict UJ.
To make sure the same dependencies_schemas can propagate in both cases, we should auto-detect prediction_context in pyfunc predict/predict_stream. If there's no context then we create an empty context without request_id, and potentially store dependencies_schemas within the context.
NB: we don't pass request_id in context for offline predict because the request_id should always be generated by backend store.

### How is this PR tested?

- [ ] Existing unit/integration tests
- [x] New unit/integration tests
- [ ] Manual tests

### Does this PR require documentation update?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

- [x] Yes (this PR will be cherry-picked and included in the next patch release)
- [ ] No (this PR will be included in the next minor release)

### CODE REVIEW

File: mlflow/pyfunc/__init__.py [hunk 1]
```diff
@@ -720,7 +721,22 @@ def _update_dependencies_schemas_in_prediction_context(self):
...
+        # with a request_id
+        if (context := _try_get_prediction_context()) is None:
+            context = Context()
+        with set_prediction_context(context):
+            yield context
+
     def predict(self, data: PyFuncInput, params: Optional[Dict[str, Any]] = None) -> PyFuncOutput:
+        with self._try_get_or_generate_prediction_context() as context:
+            self._update_dependencies_schemas_in_prediction_context(context)
+            return self._predict(data, params)
```
Comment (sunishsheth2009) [inline mlflow/pyfunc/__init__.py:737]: Do we want to create a context even if users don't have tracing related code while using pyfunc? Would these cause any empty trace info with tags being created?

Comment (serena-ruan) [reply mlflow/pyfunc/__init__.py:737]: If user doesn't trace their code, then no traces will be logged at all. I can add a test case for this :)

File: mlflow/pyfunc/__init__.py [hunk 2]
```diff
@@ -720,7 +721,22 @@ def _update_dependencies_schemas_in_prediction_context(self):
                 }
             )
 
+    @contextmanager
+    def _try_get_or_generate_prediction_context(self):
+        # set context for prediction if it's not set
+        # NB: in model serving the prediction context must be set
+        # with a request_id
+        if (context := _try_get_prediction_context()) is None:
+            context = Context()
```
Comment (harupy) [inline mlflow/pyfunc/__init__.py:730]:
```suggestion
        context = _try_get_prediction_context()) or Context()
```

File: mlflow/tracing/processor/mlflow.py [hunk 3]
```diff
@@ -92,14 +93,16 @@ def _start_trace(self, span: OTelSpan) -> TraceInfo:
                 "To avoid performance and disambiguation issues, set the experiment for "
                 "your environment using `mlflow.set_experiment()` API."
             )
-        default_tags = resolve_tags()
-        default_tags.update({TRACE_SCHEMA_VERSION_KEY: str(TRACE_SCHEMA_VERSION)})
+        tags = resolve_tags()
```
Comment (sunishsheth2009) [inline mlflow/tracing/processor/mlflow.py:96]: Just to confirm, this code change is just a pure rename?

**Review by sunishsheth2009 — APPROVED** [review]
LGTM, thank you :)

File: mlflow/tracing/processor/mlflow.py (hunk 3 above)
Comment (serena-ruan) [reply mlflow/tracing/processor/mlflow.py:96]: Yes 😆

**Pull request merged**